File: src/main/java/io/github/seggan/jyxal/runtime/RuntimeHelpers.java
package io.github.seggan.jyxal.runtime;

import io.github.seggan.jyxal.runtime.list.JyxalList;
import io.github.seggan.jyxal.runtime.math.BigComplex;
import jdk.jshell.JShell;
import jdk.jshell.Snippet;
import jdk.jshell.SnippetEvent;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class RuntimeHelpers {

    public static final Set<Character> VOWELS = Set.of('a', 'e', 'i', 'o', 'u', 'y');
    public static final LazyInit<JShell> jShell = new LazyInit<>(() -> {
        JShell shell = JShell.create();
        Runtime.getRuntime().addShutdownHook(new Thread(shell::close));
        return shell;
    });
    public static final LazyInit<Pattern> NUMBER_PATTERN = new LazyInit<>(() -> Pattern.compile("\\d+(\\.\\d+)?"));
    public static final LazyInit<Pattern> LIST_PATTERN = new LazyInit<>(() -> Pattern.compile("\\[.+(?:,(.+))*]"));

    private RuntimeHelpers() {
    }

    public static Object applyLambda(Lambda lambda, Object obj) {
        if (obj instanceof JyxalList jyxalList) {
            JyxalList list = JyxalList.create();
            for (Object o : jyxalList) {
                list.add(lambda.call(o));
            }
            return list;
        } else if (obj instanceof BigComplex bigComplex) {
            BigComplex current = BigComplex.ONE;
            JyxalList list = JyxalList.create();
            while (current.compareTo(bigComplex) <= 0) {
                list.add(lambda.call(current));
            }
            return list;
        } else {
            String s = obj.toString();
            JyxalList list = JyxalList.create();
            for (char c : s.toCharArray()) {
                list.add(lambda.call(Character.toString(c)));
            }
            return list;
        }
    }

    public static JyxalList deepCopy(JyxalList list) {
        JyxalList copy = JyxalList.create();
        for (Object obj : list) {
            if (obj instanceof JyxalList jyxalList) {
                copy.add(deepCopy(jyxalList));
            } else {
                copy.add(obj);
            }
        }
        return copy;
    }

    public static Object exec(String expr) {
        ProgramStack stack = new ProgramStack();
        for (SnippetEvent e : jShell.get().eval(jShell.get().sourceCodeAnalysis().analyzeCompletion(expr).source())) {
            if (e.status() == Snippet.Status.VALID) {
                pushExpr(stack, e.value());
            } else {
                throw new RuntimeException(e.toString());
            }
        }
        return stack.pop();
    }

    public static Object filterLambda(Lambda lambda, Object obj) {
        if (obj instanceof JyxalList jyxalList) {
            return jyxalList.filter(o -> truthValue(lambda.call(o)));
        } else if (obj instanceof BigComplex bigComplex) {
            return JyxalList.range(BigComplex.ONE, bigComplex).filter(o -> truthValue(lambda.call(o)));
        } else {
            String s = obj.toString();
            JyxalList list = JyxalList.create();
            for (char c : s.toCharArray()) {
                if (truthValue(lambda.call(Character.toString(c)))) {
                    list.add(Character.toString(c));
                }
            }
            return list;
        }
    }

    public static Iterator<Object> forify(ProgramStack stack) {
        return forify(stack.pop());
    }

    public static Iterator<Object> forify(Object obj) {
        if (obj instanceof JyxalList jyxalList) {
            return jyxalList.iterator();
        } else if (obj instanceof BigComplex bigComplex) {
            return new Iterator<>() {
                private BigComplex current = BigComplex.ONE;

                @Override
                public boolean hasNext() {
                    return current.re.compareTo(bigComplex.re) <= 0;
                }

                @Override
                public Object next() {
                    BigComplex next = current;
                    current = current.add(BigComplex.ONE);
                    return next;
                }
            };
        } else {
            String s = obj.toString();
            return new Iterator<>() {
                private int i = 0;

                @Override
                public boolean hasNext() {
                    return i < s.length();
                }

                @Override
                public Object next() {
                    return Character.toString(s.charAt(i++));
                }
            };
        }
    }

    public static int fromBaseDigitsAlphabet(CharSequence digits, String alphabet) {
        int result = 0;
        for (int i = 0; i < digits.length(); i++) {
            result = result * alphabet.length() + alphabet.indexOf(digits.charAt(i));
        }
        return result;
    }

    public static Object mapLambda(Lambda lambda, Object obj) {
        if (obj instanceof JyxalList jyxalList) {
            return jyxalList.map(lambda::call);
        } else if (obj instanceof BigComplex bigComplex) {
            return JyxalList.range(BigComplex.ONE, bigComplex).map(lambda::call);
        } else {
            String s = obj.toString();
            JyxalList list = JyxalList.create();
            for (char c : s.toCharArray()) {
                list.add(lambda.call(Character.toString(c)));
            }
            return list;
        }
    }

    public static <T extends Collection<BigComplex>> T primeFactors(BigComplex n, Supplier<T> factory) {
        T factors = factory.get();
        if (n.re.compareTo(BigDecimal.valueOf(Long.MAX_VALUE)) < 0
                && n.re.compareTo(BigDecimal.valueOf(Long.MIN_VALUE)) > 0) {
            // we can use primitives to speed this up
            long nLong = n.re.longValue();
            long sqrt = (long) Math.sqrt(nLong);
            for (long i = 2; i <= sqrt; i++) {
                if (nLong % i == 0) {
                    factors.add(BigComplex.valueOf(i));
                }
            }
            if (sqrt * sqrt == nLong) {
                factors.add(BigComplex.valueOf(sqrt));
            }
        } else {
            // we can't use primitives, so we'll have to use BigInteger
            BigInteger nBig = n.re.toBigInteger();
            BigInteger sqrt = nBig.sqrt();
            for (BigInteger i = BigInteger.valueOf(2); i.compareTo(sqrt) <= 0; i = i.add(BigInteger.ONE)) {
                if (nBig.mod(i).equals(BigInteger.ZERO)) {
                    factors.add(BigComplex.valueOf(new BigDecimal(i)));
                }
            }
            if (sqrt.multiply(sqrt).equals(nBig)) {
                factors.add(BigComplex.valueOf(new BigDecimal(sqrt)));
            }
        }

        return factors;
    }

    public static void pushExpr(ProgramStack stack, String expr) {
        if (NUMBER_PATTERN.get().matcher(expr).matches()) {
            stack.push(BigComplex.valueOf(new BigDecimal(expr)));
        } else {
            Matcher matcher = LIST_PATTERN.get().matcher(expr);
            if (matcher.matches()) {
                ProgramStack newStack = new ProgramStack();
                while (matcher.find()) {
                    pushExpr(newStack, matcher.group());
                }
                stack.push(JyxalList.create(newStack));
            } else if (expr.startsWith("\"") && expr.endsWith("\"")) {
                stack.push(expr.substring(1, expr.length() - 1));
            } else {
                stack.push(expr);
            }
        }
    }

    public static String repeatCharacters(String str, int times) {
        StringBuilder sb = new StringBuilder();
        for (char c : str.toCharArray()) {
            sb.append(String.valueOf(c).repeat(Math.max(0, times)));
        }

        return sb.toString();
    }

    public static Iterator<Object> replacementIterator(Iterator<Object> iterator, int index, Object replacement) {
        return new Iterator<>() {
            private int i = 0;

            @Override
            public boolean hasNext() {
                return iterator.hasNext();
            }

            @Override
            public Object next() {
                if (i++ == index) {
                    return replacement;
                } else {
                    return iterator.next();
                }
            }
        };
    }

    private static int slen(JyxalList first, JyxalList... rest) {
        int size = first.size();
        for (JyxalList list : rest) {
            size = Math.min(size, list.size());
        }
        return size;
    }

    public static List<BigInteger> toBaseDigits(BigInteger integer, BigInteger base) {
        List<BigInteger> result = new ArrayList<>();
        BigInteger remainder = integer;
        while (remainder.compareTo(base) >= 0) {
            BigInteger[] div = remainder.divideAndRemainder(base);
            result.add(div[1]);
            remainder = div[0];
        }
        result.add(remainder);
        Collections.reverse(result);
        return result;
    }

    public static String toBaseDigitsAlphabet(long integer, String alphabet) {
        StringBuilder sb = new StringBuilder();
        long remainder = integer;
        int base = alphabet.length();
        while (remainder >= base) {
            long div = remainder / base;
            sb.append(alphabet.charAt((int) (remainder % base)));
            remainder = div;
        }
        sb.append(alphabet.charAt((int) remainder));
        return sb.reverse().toString();
    }

    public static boolean truthValue(ProgramStack stack) {
        return truthValue(stack.pop());
    }

    public static boolean truthValue(Object obj) {
        if (obj instanceof JyxalList jyxalList) {
            return jyxalList.size() != 0;
        } else if (obj instanceof BigComplex bigComplex) {
            return !bigComplex.equals(BigComplex.ZERO);
        }

        return true;
    }

    /**
     * Unescapes a string that contains standard Java escape sequences.
     * <ul>
     * <li><strong>&#92;b &#92;f &#92;n &#92;r &#92;t &#92;" &#92;'</strong> :
     * BS, FF, NL, CR, TAB, double and single quote.</li>
     * <li><strong>&#92;X &#92;XX &#92;XXX</strong> : Octal character
     * specification (0 - 377, 0x00 - 0xFF).</li>
     * <li><strong>&#92;uXXXX</strong> : Hexadecimal based Unicode character.</li>
     * </ul>
     * <p>
     * Authored by uklimaschewski, placed in the public domain.
     *
     * @param st A string optionally containing standard java escape sequences.
     * @return The translated string.
     */
    public static String unescapeString(String st) {

        StringBuilder sb = new StringBuilder(st.length());

        for (int i = 0; i < st.length(); i++) {
            char ch = st.charAt(i);
            if (ch == '\\') {
                char nextChar = (i == st.length() - 1) ? '\\' : st
                        .charAt(i + 1);
                // Octal escape?
                if (nextChar >= '0' && nextChar <= '7') {
                    String code = "" + nextChar;
                    i++;
                    if ((i < st.length() - 1) && st.charAt(i + 1) >= '0'
                            && st.charAt(i + 1) <= '7') {
                        code += st.charAt(i + 1);
                        i++;
                        if ((i < st.length() - 1) && st.charAt(i + 1) >= '0'
                                && st.charAt(i + 1) <= '7') {
                            code += st.charAt(i + 1);
                            i++;
                        }
                    }
                    sb.append((char) Integer.parseInt(code, 8));
                    continue;
                }
                switch (nextChar) {
                    case '\\' -> ch = '\\';
                    case 'b' -> ch = '\b';
                    case 'f' -> ch = '\f';
                    case 'n' -> ch = '\n';
                    case 'r' -> ch = '\r';
                    case 't' -> ch = '\t';
                    case '\"' -> ch = '\"';
                    case '\'' -> ch = '\'';

                    // Hex Unicode: u????
                    case 'u' -> {
                        if (i >= st.length() - 5) {
                            ch = 'u';
                            break;
                        }
                        int code = Integer.parseInt(
                                "" + st.charAt(i + 2) + st.charAt(i + 3)
                                        + st.charAt(i + 4) + st.charAt(i + 5), 16);
                        sb.append(Character.toChars(code));
                        i += 5;
                        continue;
                    }
                }
                i++;
            }
            sb.append(ch);
        }
        return sb.toString();
    }

    public static Object vectorise(int arity, Function<ProgramStack, Object> function, ProgramStack stack) {
        // <editor-fold desc="vectorise" defaultstate="collapsed">
        switch (arity) {
            case 1 -> {
                Object obj = stack.pop();
                if (obj instanceof JyxalList jyxalList) {
                    return jyxalList.map(o -> function.apply(new ProgramStack(o)));
                }
                stack.push(obj);
            }
            case 2 -> {
                Object right = stack.pop();
                Object left = stack.pop();
                if (left instanceof JyxalList leftList) {
                    if (right instanceof JyxalList rightList) {
                        JyxalList result = JyxalList.create();
                        int size = slen(leftList, rightList);
                        for (int i = 0; i < size; i++) {
                            result.add(function.apply(new ProgramStack(leftList.get(i), rightList.get(i))));
                        }
                        return result;
                    }
                    return leftList.map(o -> function.apply(new ProgramStack(o, right)));
                } else if (right instanceof JyxalList rightList) {
                    return rightList.map(o -> function.apply(new ProgramStack(left, o)));
                }
                stack.push(left);
                stack.push(right);
            }
            case 3 -> {
                Object right = stack.pop();
                Object middle = stack.pop();
                Object left = stack.pop();
                if (left instanceof JyxalList leftList) {
                    if (middle instanceof JyxalList middleList) {
                        if (right instanceof JyxalList rightList) {
                            JyxalList result = JyxalList.create();
                            int size = slen(leftList, middleList, rightList);
                            for (int i = 0; i < size; i++) {
                                result.add(function.apply(new ProgramStack(leftList.get(i), middleList.get(i), rightList.get(i))));
                            }
                            return result;
                        }
                        JyxalList result = JyxalList.create();
                        int size = slen(leftList, middleList);
                        for (int i = 0; i < size; i++) {
                            result.add(function.apply(new ProgramStack(leftList.get(i), middleList.get(i), right)));
                        }
                        return result;
                    }
                    if (right instanceof JyxalList rightList) {
                        JyxalList result = JyxalList.create();
                        int size = slen(leftList, rightList);
                        for (int i = 0; i < size; i++) {
                            result.add(function.apply(new ProgramStack(leftList.get(i), middle, rightList.get(i))));
                        }
                        return result;
                    }
                    return leftList.map(o -> function.apply(new ProgramStack(o, middle, right)));
                } else if (middle instanceof JyxalList middleList) {
                    if (right instanceof JyxalList rightList) {
                        JyxalList result = JyxalList.create();
                        int size = slen(middleList, rightList);
                        for (int i = 0; i < size; i++) {
                            result.add(function.apply(new ProgramStack(left, middleList.get(i), rightList.get(i))));
                        }
                        return result;
                    }
                    JyxalList result = JyxalList.create();
                    int size = slen(middleList);
                    for (int i = 0; i < size; i++) {
                        result.add(function.apply(new ProgramStack(left, middleList.get(i), right)));
                    }
                    return result;
                } else if (right instanceof JyxalList rightList) {
                    JyxalList result = JyxalList.create();
                    int size = slen(rightList);
                    for (int i = 0; i < size; i++) {
                        result.add(function.apply(new ProgramStack(left, middle, rightList.get(i))));
                    }
                    return result;
                }
                stack.push(left);
                stack.push(middle);
                stack.push(right);
            }
        }

        return null;
        // </editor-fold>
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/runtime/RuntimeHelpers.kt
@file:Suppress("MemberVisibilityCanBePrivate", "unused")
@file:JvmName("RuntimeHelpers")

package io.github.seggan.jyxal.runtime

import io.github.seggan.jyxal.runtime.list.JyxalList
import io.github.seggan.jyxal.runtime.math.BigComplex
import jdk.jshell.JShell
import jdk.jshell.Snippet
import java.math.BigDecimal
import java.math.BigInteger
import java.util.function.Supplier
import java.util.regex.Matcher
import java.util.regex.Pattern
import kotlin.math.sqrt

internal val VOWELS = setOf('a', 'e', 'i', 'o', 'u', 'y')
private val jShell: JShell by lazy(LazyThreadSafetyMode.NONE) {
    val shell = JShell.create()
    Runtime.getRuntime().addShutdownHook(Thread { shell.close() })
    shell
}
private val NUMBER_PATTERN: Pattern by lazy(LazyThreadSafetyMode.NONE) { Pattern.compile("\\d+(\\.\\d+)?") }
private val LIST_PATTERN: Pattern by lazy(LazyThreadSafetyMode.NONE) { Pattern.compile("\\[.+(?:,(.+))*]") }

fun applyLambda(lambda: Lambda, obj: Any): Any {
    when (obj) {
        is JyxalList -> {
            val newList = JyxalList.create()
            for (item in obj) {
                newList.add(applyLambda(lambda, item))
            }
            return newList
        }
        is BigComplex -> {
            val current = BigComplex.ONE
            val list = JyxalList.create()
            while (current <= obj) {
                list.add(lambda.call(current))
            }
            return list
        }
        else -> {
            val s = obj.toString()
            val list = JyxalList.create()
            for (c in s) {
                list.add(lambda.call(c.toString()))
            }
            return list
        }
    }
}

fun copy(obj: Any): Any {
    return if (obj is JyxalList) {
        obj.map(::copy)
    } else obj
}

fun exec(expr: String): Any {
    val stack = ProgramStack()
    for (e in jShell.eval(jShell.sourceCodeAnalysis().analyzeCompletion(expr).source())) {
        if (e.status() == Snippet.Status.VALID) {
            stack.push(eval(e.value()))
        } else {
            throw RuntimeException(e.toString())
        }
    }
    return stack.pop()
}

fun escapeString(str: String): String {
    val sb = StringBuilder()
    for (c in str) {
        when (c) {
            '\n' -> sb.append("\\n")
            '\r' -> sb.append("\\r")
            '\t' -> sb.append("\\t")
            '\b' -> sb.append("\\b")
            '\$' -> sb.append("\\$")
            '\\' -> sb.append("\\\\")
            '"' -> sb.append("\\\"")
            else -> sb.append(c)
        }
    }
    return sb.toString()
}

fun eval(expr: String): Any {
    return if (NUMBER_PATTERN.matcher(expr).matches()) {
        BigComplex.valueOf(BigDecimal(expr))
    } else {
        val matcher: Matcher = LIST_PATTERN.matcher(expr)
        return if (matcher.matches()) {
            val list = JyxalList.create()
            while (matcher.find()) {
                list.add(eval(matcher.group(1)))
            }
            list
        } else if (expr.startsWith("\"") && expr.endsWith("\"")) {
            expr.substring(1, expr.length - 1)
        } else {
            expr
        }
    }
}

fun filterLambda(lambda: Lambda, obj: Any): Any {
    return when (obj) {
        is JyxalList -> {
            obj.filter { truthValue(lambda.call(it)) }
        }
        is BigComplex -> {
            JyxalList.range(BigComplex.ONE, obj).filter { truthValue(lambda.call(it)) }
        }
        else -> {
            val s = obj.toString()
            val list = JyxalList.create()
            for (c in s) {
                if (truthValue(lambda.call(c.toString()))) {
                    list.add(c.toString())
                }
            }
            list
        }
    }
}

fun forify(stack: ProgramStack): Iterator<Any> {
    return forify(stack.pop())
}

fun forify(obj: Any): Iterator<Any> {
    return if (obj is BigComplex) {
        return object : Iterator<Any> {
            private var current = BigComplex.ONE

            override fun hasNext(): Boolean {
                return current.re <= obj.re
            }

            override fun next(): Any {
                val next = current
                current = current.plus(BigComplex.ONE)
                return next
            }
        }
    } else {
        iterator(obj)
    }
}

fun fromBaseDigitsAlphabet(digits: CharSequence, alphabet: String): Int {
    var result = 0
    for (element in digits) {
        result = result * alphabet.length + alphabet.indexOf(element)
    }
    return result
}

fun iterator(obj: Any): Iterator<Any> {
    if (obj is JyxalList) {
        return obj.iterator()
    } else {
        val s = obj.toString()
        return object : Iterator<Any> {
            private var index = 0

            override fun hasNext(): Boolean {
                return index < s.length
            }

            override fun next(): Any {
                return s[index++].toString()
            }
        }
    }
}

fun len(obj: Any): Int {
    return if (obj is JyxalList) {
        obj.size
    } else {
        obj.toString().length
    }
}

fun listify(obj: Any): JyxalList {
    return if (obj is JyxalList) {
        obj
    } else {
        JyxalList.create(iterator(obj))
    }
}

fun mapLambda(lambda: Lambda, obj: Any): Any {
    return when (obj) {
        is JyxalList -> {
            obj.map { lambda.call(it) }
        }
        is BigComplex -> {
            JyxalList.range(BigComplex.ONE, obj).map { lambda.call(it) }
        }
        else -> {
            val s = obj.toString()
            val list = JyxalList.create()
            for (c in s) {
                list.add(lambda.call(c.toString()))
            }
            list
        }
    }
}

fun <T : MutableCollection<BigComplex>> primeFactors(n: BigComplex, factory: Supplier<T>): T {
    val factors = factory.get()
    if (n.re < BigDecimal.valueOf(Long.MAX_VALUE) && n.re > BigDecimal.valueOf(Long.MIN_VALUE)) {
        // we can use primitives to speed this up
        val nLong = n.re.toLong()
        val sqrt = sqrt(nLong.toDouble()).toLong()
        for (i in 2..sqrt) {
            if (nLong % i == 0L) {
                factors.add(BigComplex.valueOf(i))
            }
        }
        if (sqrt * sqrt == nLong) {
            factors.add(BigComplex.valueOf(sqrt))
        }
    } else {
        // we can't use primitives, so we'll have to use BigInteger
        val nBig = n.re.toBigInteger()
        val sqrt = nBig.sqrt()
        var i = BigInteger.valueOf(2)
        while (i <= sqrt) {
            if (nBig.mod(i) == BigInteger.ZERO) {
                factors.add(BigComplex.valueOf(BigDecimal(i)))
            }
            i += BigInteger.ONE
        }
        if (sqrt.multiply(sqrt) == nBig) {
            factors.add(BigComplex.valueOf(BigDecimal(sqrt)))
        }
    }
    return factors
}

fun repeatCharacters(str: String, times: Int): String {
    val sb = StringBuilder()
    for (c in str) {
        sb.append(c.toString().repeat(0.coerceAtLeast(times)))
    }
    return sb.toString()
}

fun replacementIterator(iterator: Iterator<Any>, index: Int, replacement: Any): Iterator<Any> {
    return object : Iterator<Any> {
        private var i = 0

        override fun hasNext(): Boolean {
            return iterator.hasNext()
        }

        override fun next(): Any {
            return if (i++ == index) {
                replacement
            } else {
                iterator.next()
            }
        }
    }
}

fun toBaseDigits(integer: BigInteger, base: BigInteger): List<BigInteger> {
    val ret = mutableListOf<BigInteger>()
    var n = integer
    while (n >= base) {
        val (newN, digit) = n.divideAndRemainder(base)
        ret.add(digit)
        n = newN
    }
    ret.add(n)
    return ret.reversed()
}

fun toBaseDigitsAlphabet(integer: Long, alphabet: String): String {
    val sb = StringBuilder()
    var n = integer
    val base = alphabet.length
    while (n >= base) {
        val div = n / base
        sb.append(alphabet[(n % base).toInt()])
        n = div
    }
    sb.append(alphabet[n.toInt()])
    return sb.reverse().toString()
}

fun truthValue(stack: ProgramStack): Boolean {
    return truthValue(stack.pop())
}

fun unescapeString(st: String): String {
    val sb = StringBuilder(st.length)
    var i = 0
    while (i < st.length) {
        var ch = st[i]
        if (ch == '\\') {
            val nextChar = if (i == st.length - 1) '\\' else st[i + 1]
            // Octal escape?
            if (nextChar in '0'..'7') {
                var code = "" + nextChar
                i++
                if (i < st.length - 1 && st[i + 1] >= '0' && st[i + 1] <= '7') {
                    code += st[i + 1]
                    i++
                    if (i < st.length - 1 && st[i + 1] >= '0' && st[i + 1] <= '7') {
                        code += st[i + 1]
                        i++
                    }
                }
                sb.append(code.toInt(8).toChar())
                i++
                continue
            }
            when (nextChar) {
                '\\' -> ch = '\\'
                'b' -> ch = '\b'
                'n' -> ch = '\n'
                'r' -> ch = '\r'
                't' -> ch = '\t'
                '\"' -> ch = '\"'
                '\'' -> ch = '\''
                'u' -> {
                    if (i >= st.length - 5) {
                        break
                    } else {
                        val code = ("" + st[i + 2] + st[i + 3] + st[i + 4] + st[i + 5]).toInt(16)
                        sb.append(Character.toChars(code))
                        i += 5
                        continue
                    }
                }
            }
            i++
        }
        sb.append(ch)
        i++
    }
    return sb.toString()
}


fun truthValue(obj: Any): Boolean {
    if (obj is JyxalList) {
        return obj.isNotEmpty()
    } else if (obj is BigComplex) {
        return obj != BigComplex.ZERO
    }

    return true
}

fun vectorise(arity: Int, function: (ProgramStack) -> Any, stack: ProgramStack): Any? {
    when (arity) {
        1 -> {
            val obj = stack.pop()
            if (obj is JyxalList) {
                return obj.map { function(ProgramStack(it)) }
            }
            stack.push(obj)
        }
        2 -> {
            val right = stack.pop()
            val left = stack.pop()
            if (left is JyxalList) {
                if (right is JyxalList) {
                    return left.zip(right) { a, b -> function(ProgramStack(a, b)) }
                }
                return left.map { function(ProgramStack(it, right)) }
            } else if (right is JyxalList) {
                return right.map { function(ProgramStack(left, it)) }
            }
            stack.push(left)
            stack.push(right)
        }
        3 -> {
            val right = stack.pop()
            val middle = stack.pop()
            val left = stack.pop()
            if (left is JyxalList) {
                if (middle is JyxalList) {
                    if (right is JyxalList) {
                        return left.zip(middle).zip(right) { a, b ->
                            a as JyxalList
                            function(ProgramStack(a[0], a[1], b))
                        }
                    }
                    return left.zip(middle) { a, b -> function(ProgramStack(a, b, right)) }
                } else if (right is JyxalList) {
                    return left.zip(right) { a, b -> function(ProgramStack(a, middle, b)) }
                }
                return left.map { function(ProgramStack(it, middle, right)) }
            } else if (middle is JyxalList) {
                if (right is JyxalList) {
                    return middle.zip(right) { a, b -> function(ProgramStack(left, a, b)) }
                }
                return middle.map { function(ProgramStack(left, it, right)) }
            } else if (right is JyxalList) {
                return right.map { function(ProgramStack(left, middle, it)) }
            }
            stack.push(left)
            stack.push(middle)
            stack.push(right)
        }
        else -> throw RuntimeException("Vectorise arity $arity not supported")
    }
    return null
}
